website/static/js/registrationUtils.js

```diff
@@ -0,0 +1,764 @@
+        });
+    }
+};
+/**
+ * Shows/hides the Question example
+ **/
+Question.prototype.toggleExample = function() {
+    this.showExample(!this.showExample());
+};
+/**
+ * Shows/hides the Question uploader
+ **/
+Question.prototype.toggleUploader = function() {
+    this.showUploader(!this.showUploader());
+};
+
+/**
+ * Model for MetaSchema instances
+ *
+ * @param {Object} params: instantiation values
+ * @param {String} params.schema_name
+ * @param {Integer} params.schema_version
+ * @param {String} params.title: display title of schema
+ * @param {Schema} params.schema
+ *
+ * @type Schema
+ * @property {String} title
+ * @property {Integer} version
+ * @property {String} description
+ * @property {String[]} fulfills: array of requirements/goals that this schema fulfills
+ * @property {Page[]} pages
+ *
+ * @type Page
+ * @property {String} id
+ * @property {String} title
+ * @property {Question[]} questions
+ **/
+var MetaSchema = function(params) {
+    var self = this;
+
+    self.name = params.schema_name;
+    self.version = params.schema_version;
+    self.title = params.title || params.schema.title;
+    self.schema = params.schema || {};
+    self.id = [self.name, self.version].join('_');
+
+    self.requiresApproval = params.requires_approval || false;
+    self.fulfills = params.fulfills || [];
+    self.messages = params.messages || {};
+
+    $.each(self.schema.pages, function(i, page) {
+        var mapped = {};
+        $.each(page.questions, function(qid, question) {
+            mapped[qid] = new Question(question, qid);
+        });
+        self.schema.pages[i].questions = mapped;
+    });
+};
+/**
+ * @returns {Question[]} a flat list of the schema's questions
+ **/
+MetaSchema.prototype.flatQuestions = function() {
+    var self = this;
+
+    var flat = [];
+
+    $.each(self.schema.pages, function(i, page) {
+        $.each(page.questions, function(qid, question) {
+            flat.push(question);
+        });
+    });
+    return flat;
+};
+
+/**
+ * @class Draft
+ * Model for DraftRegistrations
+ *
+ * @param {Object} params
+ * @param {String} params.pk: primary key of Draft
+ * @param {Object} params.registration_schema: data to be passed to MetaSchema constructor
+ * @param {Object} params.registration_metadata: saved Draft data
+ * @param {User} params.initiator
+ * @param {Date} params.initiated
+ * @param {Boolean} params.is_pending_review
+ * @param {Boolean} params.is_approved
+ * @param {Date} params.updated
+ * @param {Object} params.urls
+ * @param {String} params.urls.edit
+ * @param {String} params.urls.before_register
+ * @param {String} params.urls.register
+ * @property {Float} completion: percent completion of schema
+ **/
+var Draft = function(params, metaSchema) {
+    var self = this;
+
+    self.pk = params.pk;
+    self.metaSchema = metaSchema || new MetaSchema(params.registration_schema);
+    self.schema = ko.pureComputed(function() {
+        return self.metaSchema.schema;
```

Comment:
This [computed observable may not be necessary](https://github.com/CenterForOpenScience/osf.io/pull/3476/files#r39556712).